src/LinuxTracing/ThreadStateManager.h

```diff
@@ -60,10 +62,20 @@ class ThreadStateManager {
 
  private:
   struct OpenState {
-    OpenState(orbit_grpc_protos::ThreadStateSlice::ThreadState state, uint64_t begin_timestamp_ns)
-        : state{state}, begin_timestamp_ns{begin_timestamp_ns} {}
+    OpenState(orbit_grpc_protos::ThreadStateSlice::ThreadState state, uint64_t begin_timestamp_ns,
+              pid_t was_blocked_by_thread, pid_t was_blocked_by_process)
+        : state{state},
+          begin_timestamp_ns{begin_timestamp_ns},
+          was_blocked_by_thread{was_blocked_by_thread},
+          was_blocked_by_process{was_blocked_by_process} {}
     orbit_grpc_protos::ThreadStateSlice::ThreadState state;
     uint64_t begin_timestamp_ns;
+    // The next two fields are optional, we use them to transfer info between interruptible sleep
+    // state (gray) and runnable state (blue) so we can know which tid and pid caused the thread to
+    // turn into a runnable state if you don't need them just put zeros when using the OpenState
+    // struct.
+    pid_t was_blocked_by_thread = 0u;
```

Comment:
this 0u should be a pre defined constant. I would use the -1 defined in OrbitBase/ThreadConstants.h
```
// Cross-platform invalid thread id.
static constexpr uint32_t kInvalidThreadId = -1;  // 0xffffffff

// Cross-platform invalid process id.
static constexpr uint32_t kInvalidProcessId = -1;  // 0xffffffff
```

Reply:
Oh that's neat. I thought since `pid_t` was an unsigned int, it wouldn't take -1. 

I'm not sure if I wanna do this because the tracepoint returns zero if it can't find the pid (which sometimes can happen), I would like to leave it 0 so that we don't have to handle two different values for similar cases. We'll discuss this tomorrow.

Reply:
`pid_t` is signed. You can decide to use the constants mentioned by @karupayun or use -1 directly, but Linux and this module generally uses -1 to indicate a missing pid/tid. This is because 0 as a pid/tid is valid on Linux and refers to the scheduler. So I think we should switch to -1 across this whole change.

Reply:
I think here are two cases two consider:
1. You get 0 from perf_event_open/the kernel
2. You have a ThreadState Open from an event where there is no valid value for `was_blocked_by...`.

I suggest to use `std::optional` here and have "2." encoded as `std::nullopt`. For handling 1. differently in the UI, I'd suggest to add a constant in ThreadConstants.h.

Reply:
First of all, `pid_t` is signed. You can decide to use the constants mentioned by @karupayun or use -1 directly, but Linux and this module generally uses -1 to indicate a missing pid/tid. This is because 0 as a pid/tid is valid on Linux and refers to the scheduler.

Now, what does it mean that "the tracepoint returns zero if it can't find the pid"? Is it only the pid or also the tid? Is it because there is no waker and the kernel considers the waker the scheduler?

I would avoid using optional here and I would avoid adding even more special values to `ThreadConstants.h`. -1 is fine as "invalid" and 0 is fine as the scheduler, as those are the Linux values.

Reply:
> Oh that's neat. I thought since pid_t was an unsigned int, it wouldn't take -1.
There is a comment explaining that the value is 0xffffffff in this case. It will be converted because that -1 is signed.

> This is because 0 as a pid/tid is valid on Linux and refers to the scheduler.
I was worried about this. Thanks Davide, I wasn't sure.